Content for native testing framework section of the Leo-lang docs (#489)
ProvableHQ/leo-docs-source: documentation/testing/01_test_framework.md

sidebar_label: Testing
---

## Introduction

> The Leo testing frameworks enables developers to validate their Leo program logic by writing unit and integration tests. Tests are written in Leo and are located in a `/tests` subdirectory of the main Leo project directory.

Author: fixed:[a830dd0](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/a830dd0e3116f07bce4cb00eadff89c327a8cfbd)

```bash
example_program
├── build
│   ├── imports
│   │   └── test_example_program.aleo
│   ├── main.aleo
│   └── program.json
├── outputs
├── src
│   └── main.leo
├── tests
│   └── test_example_program.leo
└── program.json
```
The test file is a Leo program that imports the program in `main.leo`.  The test functions will all be annotated with `@test` above the function declaration. 

> This tutorial will use an example program which can be found in the [example's repository](https://github.com/ProvableHQ/leo-examples/tree/main/tests).  

Author: Link won't work until the Leo/examples PR is merged.

Author: [b9ffc53](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/b9ffc532101f9207b5585bb45ae6db23635c7aff)

:::info
Developers can add multiple `leo` files to the test directory but must ensure that the name of the test file matches the program name.  For example, if the name of the test file is `test_example_program.leo`, the program name must be `test_example_program.aleo`.
:::

## Testing transition function logic.

The `example_program.leo` program contains a transition function which returns the sum of two `u32` inputs.

```Leo
transition simple_addition(public a: u32, b: u32) -> u32 {
    let c: u32 = a + b;
    return c;
}
```

The `test_example_program.leo` contains two tests to ensure that the transition logic returns a correct output and fails when the output does not match the sum of the input values.
```Leo
@test
transition test_simple_addition() {
    let result: u32 = example_program.aleo/simple_addition(2u32, 3u32);
    assert_eq(result, 5u32);
}
```

> The `@should_fail` annotation should be added after the `@test` annotation for tests that are expected to fail.

Reviewer: I believe the order of the annotations does not matter. Maybe we should say "along with" instead of "after"

Author: Even if the ordering is not a strict requriement, would we want to encourage devs to follow a convention here?

Reviewer: Yup that makes sense.

```Leo
@test
@should_fail
transition test_simple_addition_fail() {
    let result: u32 = example_program.aleo/simple_addition(2u32, 3u32);
    assert_eq(result, 3u32);
}
```

> Developers can test that Record and struct fields match their expected values.

Reviewer: I feel like this is added out of context? Or are we introducing a brand new example here?

Author: I split the `Testing transition functions` section into two parts.
[022c5a](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/3022c5afef4cf6df91604be2166a451eb9f899c3)

```Leo
@test
transition test_record_maker() {
    let r: example_program.aleo/Example = example_program.aleo/mint_record(0field);
    assert_eq(r.x, 0field);
}
```

:::info

> Each test file is required to have at least one transition function.

Author: @mikebenfield We briefly spoke about expanding this blurb... not sure how much we want to add here since the requirements may change in the near term..

:::

## Using interpreted tests for modelling on-chain state

> While the testing framework cannot access on-chain state from either Testnet or Mainnet, developers can simulate on-chain state using interpreted tests.  Developers are able to await Futures and update mappings using interpreted tests.  When using interpreted tests, the `transition` or `function` keyword is replaced with the `script` keyword.

Reviewer: ```suggestion
While the testing framework cannot access on-chain state from either `testnet` or `mainnet`, developers can simulate on-chain state using interpreted tests. Developers are able to await `Future`s and update mappings using interpreted tests. When using interpreted tests, the `transition` or `function` keyword is replaced with the `script` keyword.
```

Author: fixed [951b0db](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/951b0dbf32dc5677254d5f73377ccd11d80bda61)

```Leo
@test
script test_async() {
    const VAL: field = 12field;
    let fut: Future = example_program.aleo/set_mapping(VAL);
    fut.await();
    assert_eq(Mapping::get(example_program.aleo/map, 0field), VAL);

    let rand_val: field = ChaCha::rand_field();
    Mapping::set(example_program.aleo/map, VAL, rand_val);
    let value: field = Mapping::get(example_program.aleo/map, VAL);
    assert_eq(value, rand_val);
}
```

:::info

> While non-async logic may be tested using an interpreted tests, all async functionality must be tested within interpreted tests.

Author: ie transition functions that mint records, perform arbitrary computations, etc can be used in either compiled or interpreted tests..but asynchronous functions or code logic that updates mappings can only be used in interpreted tests.

:::

## Running tests

> Invoking the `leo test` command will run all of the compiled and interpreted tests. Developer may optionally select individual tests by supplying the function test name as follows:

Author: Good to know!  
updated:  [b8d2448](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/b8d24488556a91581f6866c9700e82b21bd1eafa)

:::info
Developers can add multiple `leo` files to the test directory but must ensure that the name of the test file matches the test program name.  For example, if the name of the test file is `test_example_program.leo`, the test program name must be `test_example_program.aleo`.
:::

## Testing transition functions.

### Function logic

The `example_program.leo` program contains a transition function which returns the sum of two `u32` inputs.

```Leo
transition simple_addition(public a: u32, b: u32) -> u32 {
    let c: u32 = a + b;
    return c;
}
```

The `test_example_program.leo` contains two tests to ensure that the transition logic returns a correct output and fails when the output does not match the sum of the input values.
```Leo
@test
transition test_simple_addition() {
    let result: u32 = example_program.aleo/simple_addition(2u32, 3u32);
    assert_eq(result, 5u32);
}
```

The `@should_fail` annotation should be added after the `@test` annotation for tests that are expected to fail.
```Leo
@test
@should_fail
transition test_simple_addition_fail() {
    let result: u32 = example_program.aleo/simple_addition(2u32, 3u32);
    assert_eq(result, 3u32);
}
```

## Leo types

> Developers can test that Record and struct fields match their expected values.  In `example_program.leo`, a Record is minted transition function shown here:

Reviewer 2: ```suggestion
Developers can test that record and struct fields match their expected values.  In `example_program.leo`, a record is minted transition function shown here:
```
Or enclose with "`"

Author: done [4c388bf](https://github.com/ProvableHQ/leo-docs-source/pull/489/commits/4c388bf63cb6f7fdde5a7a729e8851159d2a9609)

```Leo
record Example {
    owner: address,
    x: field,
}

transition mint_record(x: field) -> Example {
    return Example {
        owner: self.signer,
        x,
    };
}
```

The corresponding test in `test_example_program.leo` checks that the Record field contains the correct value:

```Leo
@test
transition test_record_maker() {
    let r: example_program.aleo/Example = example_program.aleo/mint_record(0field);
    assert_eq(r.x, 0field);
}
```

:::info
Each test file is required to have at least one transition function.
:::

## Using interpreted tests for modelling on-chain state

> While the testing framework cannot access on-chain state from either `testnet` or `mainnet`, developers can simulate on-chain state using interpreted tests. Developers are able to await `Future`s and update mappings using interpreted tests. When using interpreted tests, the `transition` or `function` keyword is replaced with the `script` keyword.

Reviewer 2: ```suggestion
While the testing framework cannot access on-chain state from either `testnet` or `mainnet`, developers can simulate on-chain state in `script`s. A script is interpreted Leo code in which developers are able to await `Future`s and update mappings using interpreted tests. When using interpreted tests, the `transition` or `function` keyword is replaced with the `script` keyword.
```
